[WIP] [act] Allow PFBuilder cases without type and only predicate

jypma opened the pull request:
This is especially useful when you have a typed `PFBuilder`, e.g.
`PFBuilder<Event,Void>` and you just want to say:

```
b.match(Event::isMyType, evt -> doStuff())
```

@ktoso follow-up of https://github.com/ktoso/akka/pull/9 against master (to separate it from the Java HTTP DSL effort), and not changing `ReceiveBuilder`. It should be safe to also update `UnitPFBuilder` though.

johanandren commented:
@ktoso s comment https://github.com/ktoso/akka/pull/9#issuecomment-211324169 still holds true, or is this different?

jypma commented:
That comment was about `ReceiveBuilder`, where these changes indeed can't be applied, since it always has `Object` as input type. This PR is only about plain `PFBuilder` (and possibly `UnitPFBuilder`).

Actually, now that I've read through the code again, there's been a misunderstanding on my part about the current type safety of `PFBuilder`. It (and `UnitPFBuilder`) doesn't actually use its `<I>` type argument to type-check any of its `match` methods. Here's the current code:
```
class PFBuilder<I, R> {
  public <P> PFBuilder<I, R> match(final Class<? extends P> type, FI.Apply<? extends P, R> apply) {
    ...
  }
}
```
This allows a match branch on a `PFBuilder<I,R>`, which results in a `PartialFunction<I,R>`, to be given a match with an unrelated type `P`, that will then get invoked with instances of `I`.

In other words, the following already compiles:
```
new PFBuilder<String,Integer>()
      .match(String.class, s -> Integer.parseInt(s))
      .match(UUID.class, uuid -> 42)
      .build();
```
It won't blow up run-time (the `UUID` branch is checked because of the class argument), but allowing it compile-time is a bit nonsensical.

I'm thinking about a more structural change, involving
- Ensuring type safety at runtime as long as compiler generics are enforced
- Making predicate and transformation functions properly type co/contra variant at the call site (like scala `Function` is `[-T,+R]`)

That would result in something like:
```
class PFBuilder<I, R> {
  public <P extends I> PFBuilder<I, R> match(
     final Class<? extends P> type, 
     FI.TypedPredicate<? super P> predicate,
     FI.Apply<? super P, ? extends R> apply) {
    ...
  }

  public <P extends I> PFBuilder<I, R> match(
     final Class<? extends P> type, 
     FI.Apply<? super P, ? extends R> apply) {
    ...
  }

  public PFBuilder<I, R> match(
     FI.TypedPredicate<? super I> predicate,
     FI.Apply<? super I, ? extends R> apply) {
    ...
  }

}
```

I'll update this PR with some actual test to show what I mean.

johanandren commented:
Ah, thanks, didn't catch the `PFBuilder` vs `ReceiveBuilder` difference in scope.

johanandren commented:
OK TO TEST

drewhk commented:
LGTM, useful addition.

jypma commented:
Please don't merge yet, I'm doing some larger changes that I described earlier, and actually discovered a few unsafe scenarios in the current code. I'll have something up soon.

jypma committed:
[act] Allow PFBuilder cases without type and only predicate

This is especially useful when you have a typed PFBuilder, e.g.
PFBuilder<Event,Void> and you just want to say:

    b.match(Event::isMyType, evt -> doStuff())

This also fixes a few unsafe cases, and tightens up the type system
generally. There are minor changes in how one does matching of generic
types from now on.

jypma commented:
Updated. It turns out there were quite a few unsafe problems with the current PFBuilder:

### type / predicate safety
```
  public <P> PFBuilder<I, R> match(final Class<? extends P> type,
                                   final FI.TypedPredicate<? extends P> predicate,
                                   final FI.Apply<? extends P, R> apply) {
```
Type and predicate don't have to be related, so this will compile fine, but throw at runtime:
```
    new PFBuilder<Object,String>().match(Integer.class, (String s) -> s.isEmpty(), Object::toString).build().apply(123);
```
The only clean solution I could come up with was to tighten up the types in this match variant (making it no longer cover the generic solution for which `? extends P` was introduced), and making a new variant that relies on run-time reification. A bit hackish, but safe; alternatives are welcome.

### equals type safety
```
  public <P> PFBuilder<I, R> matchEquals(final P object,
                                         final FI.Apply<P, R> apply) {
```
This assumes that for an object `a`, if `object.equals(a)` returns true, that `a` must in fact be an instance of `P`. That's not guaranteed at all:
```
    class VeryEqualClass { public boolean equals(Object o) { return true; } };
    VeryEqualClass e = new VeryEqualClass();
    new PFBuilder<Object,String>().matchEquals(e, t -> "").build().apply(123);
```
compiles fine but throws at runtime.

Both of these are addressed by this PR now. And (1) still introduced the new method I was actually looking for in the first place :)

jypma commented:
Right, so although it compiles fine locally, the CI server doesn't like 
```
[error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-actor/src/main/java/akka/japi/pf/Reflect.java:66: package jdk.internal.org.objectweb.asm does not exist
```
I don't particularly like the internal jdk direction either, so if anyone has some tips, that's welcome.

ktoso commented on akka-actor/src/main/java/akka/japi/pf/Reflect.java:
```diff
@@ -0,0 +1,134 @@
+            try {
+                if (methodRef[1].startsWith("lambda")) {
+                    // This is an inline lambda expression, e.g. (String s) -> s.isEmpty()
+                    String argumentType = jdk.internal.org.objectweb.asm.Type.getArgumentTypes(methodRef[2])[argumentIndex].getClassName();
```
hm, are you sure this isn't reachable using plain reflection?
// please don't use internal JDK api, we won't be able to upgrade to JDK9 with such thingies :)

jypma replied:
It seems compiled type information for lambdas isn't available runtime otherwise.

The only alternative I see is providing a `.matchUnsafe(Class<A>, Function<B,C>)` that allows one to match one class to an apply of a totally different type. That's what's needed unfortunately when dealing with generic types in Java, and why the unrelated types were originally introduced in https://github.com/akka/akka/issues/16245 . It seems to have been overlooked back then that by saying `<P> Type1<? extends P>, Type2<? extends P>`, the compiler will happily drop `P` all the way down to `Object`.

So, if using internal JDK is ill-advised, it's choosing between
1. Having no compile safety at all on these methods (as it is currently)
2. Have two methods, one which is compile-safe (but can't be used with generic types), and another which isn't compile-safe and does blind casting underneath (as it is currently)

I'd vote for (2), but we'd have to find some not-too-frightening names.

patriknw commented:
@jypma sorry if this stalled on our end. I agree that two separate methods sounds right.

patriknw commented:
@jypma what is the status here?

jypma commented:
It's been dormant for a while; I still would like to pick this up at some point, since I feel the current `PFBuilder` is too limited and too unsafe for "general" partial function building (i.e. outside of plain actor type-less receive loops).

However, I don't expect to be able to put time into it in the short term. If you can agree to leave the "Experimental" flag on the class for the time being, then we can for the moment close the PR, until I (or someone else) comes back to it.

Maybe extract my earlier comment about type unsafety into a ticket?

jypma commented:
Note to self: `PFBuilder` (or a class close by) should also get static `.narrow` and `.widen` operations so `PartialFunction`s can be legally narrowed and widened in pure Java code, according to the (co)variance of their in and outputs.

jypma commented:
Actually, that could perhaps be done nicely by making `.build()` generic.

guidomedina commented:
Not sure if it is related but I have found another example where this fails, the commented lines failed at compile time:
```
receive(ReceiveBuilder.
       match(UpdateSuccess.class, this::processUpdateSuccess).
       //match(UpdateFailure.class, this::processUpdateFailure).
       //match(GetSuccess.class, this::processGetSuccess).
       //match(GetFailure.class, this::processGetFailure).
       matchAny(supervisor::processOthers).
       build())

  private void processUpdateSuccess(UpdateSuccess<ORSet<Request<?>>> success) {
    ...
  }

  private void processUpdateFailure(UpdateFailure<ORSet<Request<?>>> failure) {
    ...
  }

  private void processGetSuccess(GetSuccess<ORSet<Request<?>>> success) throws Exception {
    ...
  }

  private void processGetFailure(GetFailure<ORSet<Request<?>>> failure) {
    ...
  }
```

jypma commented:
Note to self: get some inspiration from (javaslang's matcher DSL)[http://www.javaslang.io/javaslang-docs/#_pattern_matching]

patriknw commented:
@jypma Shall we close this until it is picked up again? Is there a corresponding issue?

jypma commented:
@patriknw let's say that if we can prove that https://github.com/akka/akka/pull/22003 can in fact allow receive loops to be written in e.g. javaslang, we don't need this PR anymore.

patriknw commented:
sounds good, closing then
